In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [72]:
data = pd.read_csv("../data/clean.csv", index_col=0)
data

group  region  01D2Z36  0319-6G20  0319-7L14  27F-1492R  37-13  \
SRR4457864       0       0      0.0        0.0        0.0        0.0    0.0   
SRR4457873       0       0      0.0        0.0        0.0        0.0    0.0   
SRR4457876       0       0      0.0        0.0        0.0        0.0    0.0   
SRR4457877       0       0      0.0        0.0        0.0        0.0    0.0   
SRR4457878       0       0      0.0        0.0        0.0        0.0    0.0   
...            ...     ...      ...        ...        ...        ...    ...   
SRR24874401      1       1      0.0        0.0        0.0        0.0    0.0   
SRR24874402      1       1      0.0        0.0        0.0        0.0    0.0   
SRR24874404      1       1      0.0        0.0        0.0        0.0    0.0   
SRR24874405      1       1      0.0        0.0        0.0        0.0    0.0   
SRR24874406      1       1      0.0        0.0        0.0        0.0    0.0   

             67-14   A2  Abiotrophia  ...  Xanthomonas  Xenorhabdus  \
SRR4457864     0.0  0.0     0.000000  ...          0.0          0.0   
SRR4457873     0.0  0.0     0.000000  ...          0.0          0.0   
SRR4457876     0.0  0.0     0.000000  ...          0.0          0.0   
SRR4457877     0.0  0.0     0.000000  ...          0.0          0.0   
SRR4457878     0.0  0.0     0.000000  ...          0.0          0.0   
...            ...  ...          ...  ...          ...          ...   
SRR24874401    0.0  0.0     0.009899  ...          0.0          0.0   
SRR24874402    0.0  0.0     0.000000  ...          0.0          0.0   
SRR24874404    0.0  0.0     0.000000  ...          0.0          0.0   
SRR24874405    0.0  0.0     0.050805  ...          0.0          0.0   
SRR24874406    0.0  0.0     0.000000  ...          0.0          0.0   

             Xylanibacillus  Yersinia  Youngiibacter  Zavarzinia  Zoogloea  \
SRR4457864              0.0       0.0            0.0         0.0       0.0   
SRR4457873              0.0       0.0            0.0         0.0       0.0   
SRR4457876              0.0       0.0            0.0         0.0       0.0   
SRR4457877              0.0       0.0            0.0         0.0       0.0   
SRR4457878              0.0       0.0            0.0         0.0       0.0   
...                     ...       ...            ...         ...       ...   
SRR24874401             0.0       0.0            0.0         0.0       0.0   
SRR24874402             0.0       0.0            0.0         0.0       0.0   
SRR24874404             0.0       0.0            0.0         0.0       0.0   
SRR24874405             0.0       0.0            0.0         0.0       0.0   
SRR24874406             0.0       0.0            0.0         0.0       0.0   

             ZOR0006  Unclassified  study_site  
SRR4457864       0.0      3.640935           1  
SRR4457873       0.0      7.630561           1  
SRR4457876       0.0      7.448002           1  
SRR4457877       0.0      6.459400           1  
SRR4457878       0.0      7.031991           1  
...              ...           ...         ...  
SRR24874401      0.0     25.423813          14  
SRR24874402      0.0      3.704500          14  
SRR24874404      0.0      1.894488          14  
SRR24874405      0.0      0.111772          14  
SRR24874406      0.0      2.553967          14  

[948 rows x 1130 columns]

In [20]:
# for every study, get the number of nonzero and zero features
for i in range(1, 15):
    study = data[data['study_site'] == i]
    print(f'For study {i}, there are {(study.sum() == 0).sum()} trivial features, and {(study.sum() != 0).sum()} non-trivial features.')

For study 1, there are 888 trivial features, and 243 non-trivial features.
For study 2, there are 1017 trivial features, and 114 non-trivial features.
For study 3, there are 858 trivial features, and 273 non-trivial features.
For study 4, there are 810 trivial features, and 321 non-trivial features.
For study 5, there are 772 trivial features, and 359 non-trivial features.
For study 6, there are 923 trivial features, and 208 non-trivial features.
For study 7, there are 931 trivial features, and 200 non-trivial features.
For study 8, there are 982 trivial features, and 149 non-trivial features.
For study 9, there are 887 trivial features, and 244 non-trivial features.
For study 10, there are 946 trivial features, and 185 non-trivial features.
For study 11, there are 516 trivial features, and 615 non-trivial features.
For study 12, there are 952 trivial features, and 179 non-trivial features.
For study 13, there are 988 trivial features, and 143 non-trivial features.
For study 14, there 

In [47]:
# check how many features are unique to a study
data_by_site = data.groupby('study_site').sum()
(data_by_site.iloc[:, 2:] != 0).sum().value_counts()

1     644
2     128
3      78
11     51
4      31
14     29
0      28
5      20
10     19
8      19
7      19
12     18
6      17
13     16
9      10
Name: count, dtype: int64

There are 28 features where no study contained nonzero values for its abundance. 

There are 644 features where only one study contained nonzero values for its abundance.

In [66]:
# filter out rare OTUs: < 0.01 (values are in percentages)
data_16s = data.drop(columns=['group', 'region', 'study_site'])
rare_OTU_threshold = 1
rare_OTUs = data_16s.columns[(data_16s <= 1).sum() == data_16s.shape[0]]

In [67]:
filter_rare = data_16s.drop(columns=rare_OTUs)
filter_rare

Acetatifactor  Acetivibrio  Acidaminococcus  Acinetobacter  \
Unnamed: 0                                                                
SRR4457864             0.0          0.0         0.000000            0.0   
SRR4457873             0.0          0.0         0.000000            0.0   
SRR4457876             0.0          0.0         0.000000            0.0   
SRR4457877             0.0          0.0         0.000000            0.0   
SRR4457878             0.0          0.0         0.000000            0.0   
...                    ...          ...              ...            ...   
SRR24874401            0.0          0.0         0.000000            0.0   
SRR24874402            0.0          0.0         0.000000            0.0   
SRR24874404            0.0          0.0         0.000000            0.0   
SRR24874405            0.0          0.0         0.003387            0.0   
SRR24874406            0.0          0.0         0.000000            0.0   

             Actinomyces  Acutalibacter  Adlercreutzia  Aeromicrobium  \
Unnamed: 0                                                              
SRR4457864      0.026594            0.0       0.033847            0.0   
SRR4457873      0.016925            0.0       0.000000            0.0   
SRR4457876      0.000000            0.0       0.000000            0.0   
SRR4457877      0.019861            0.0       0.000000            0.0   
SRR4457878      0.037431            0.0       0.034935            0.0   
...                  ...            ...            ...            ...   
SRR24874401     0.262331            0.0       0.000000            0.0   
SRR24874402     0.035169            0.0       0.000000            0.0   
SRR24874404     0.474964            0.0       0.155638            0.0   
SRR24874405     0.279429            0.0       0.000000            0.0   
SRR24874406     2.364003            0.0       0.000000            0.0   

             Agathobacter  Agathobaculum  ...   UCG-003  UCG-004   UCG-005  \
Unnamed: 0                                ...                                
SRR4457864       1.716510            0.0  ...  0.055605      0.0  0.834079   
SRR4457873       2.676499            0.0  ...  0.198259      0.0  0.041103   
SRR4457876       1.985162            0.0  ...  0.192395      0.0  0.033523   
SRR4457877       2.467344            0.0  ...  0.465969      0.0  0.200137   
SRR4457878       1.247692            0.0  ...  0.416729      0.0  1.766732   
...                   ...            ...  ...       ...      ...       ...   
SRR24874401      0.000000            0.0  ...  0.274705      0.0  0.009899   
SRR24874402      6.541490            0.0  ...  0.000000      0.0  0.000000   
SRR24874404      0.614501            0.0  ...  0.000000      0.0  0.037568   
SRR24874405      0.110078            0.0  ...  0.000000      0.0  0.000000   
SRR24874406      3.617001            0.0  ...  0.011513      0.0  0.285906   

              UCG-009   UCG-010  Varibaculum  Veillonella  Weissella  \
Unnamed: 0                                                             
SRR4457864   0.000000  0.000000          0.0     0.000000        0.0   
SRR4457873   0.000000  0.009671          0.0     0.043520        0.0   
SRR4457876   0.000000  0.000000          0.0     0.119518        0.0   
SRR4457877   0.000000  0.000000          0.0     0.000000        0.0   
SRR4457878   0.064880  0.137246          0.0     0.000000        0.0   
...               ...       ...          ...          ...        ...   
SRR24874401  0.000000  0.000000          0.0     0.064345        0.0   
SRR24874402  0.000000  0.000000          0.0     0.189523        0.0   
SRR24874404  0.000000  0.000000          0.0     0.110020        0.0   
SRR24874405  0.000000  0.000000          0.0     0.450473        0.0   
SRR24874406  0.007675  0.078672          0.0     0.028783        0.0   

             Xanthomonas  Unclassified  
Unnamed: 0                              
SRR4457864           0.0      3.640935  
SRR4457873  

In [78]:
# check sparsity of resulting matrix
(filter_rare == 0).mean().mean()

np.float64(0.7596669130555299)

The filtered matrix is 76% sparse. 

In [68]:
covariates = data[['group', 'region', 'study_site']]
covariates

group  region  study_site
Unnamed: 0                            
SRR4457864       0       0           1
SRR4457873       0       0           1
SRR4457876       0       0           1
SRR4457877       0       0           1
SRR4457878       0       0           1
...            ...     ...         ...
SRR24874401      1       1          14
SRR24874402      1       1          14
SRR24874404      1       1          14
SRR24874405      1       1          14
SRR24874406      1       1          14

[948 rows x 3 columns]

In [70]:
# final clean dataset
pd.concat([covariates, filter_rare], axis=1)

group  region  study_site  Acetatifactor  Acetivibrio  \
Unnamed: 0                                                           
SRR4457864       0       0           1            0.0          0.0   
SRR4457873       0       0           1            0.0          0.0   
SRR4457876       0       0           1            0.0          0.0   
SRR4457877       0       0           1            0.0          0.0   
SRR4457878       0       0           1            0.0          0.0   
...            ...     ...         ...            ...          ...   
SRR24874401      1       1          14            0.0          0.0   
SRR24874402      1       1          14            0.0          0.0   
SRR24874404      1       1          14            0.0          0.0   
SRR24874405      1       1          14            0.0          0.0   
SRR24874406      1       1          14            0.0          0.0   

             Acidaminococcus  Acinetobacter  Actinomyces  Acutalibacter  \
Unnamed: 0                                                                
SRR4457864          0.000000            0.0     0.026594            0.0   
SRR4457873          0.000000            0.0     0.016925            0.0   
SRR4457876          0.000000            0.0     0.000000            0.0   
SRR4457877          0.000000            0.0     0.019861            0.0   
SRR4457878          0.000000            0.0     0.037431            0.0   
...                      ...            ...          ...            ...   
SRR24874401         0.000000            0.0     0.262331            0.0   
SRR24874402         0.000000            0.0     0.035169            0.0   
SRR24874404         0.000000            0.0     0.474964            0.0   
SRR24874405         0.003387            0.0     0.279429            0.0   
SRR24874406         0.000000            0.0     2.364003            0.0   

             Adlercreutzia  ...   UCG-003  UCG-004   UCG-005   UCG-009  \
Unnamed: 0                  ...                                          
SRR4457864        0.033847  ...  0.055605      0.0  0.834079  0.000000   
SRR4457873        0.000000  ...  0.198259      0.0  0.041103  0.000000   
SRR4457876        0.000000  ...  0.192395      0.0  0.033523  0.000000   
SRR4457877        0.000000  ...  0.465969      0.0  0.200137  0.000000   
SRR4457878        0.034935  ...  0.416729      0.0  1.766732  0.064880   
...                    ...  ...       ...      ...       ...       ...   
SRR24874401       0.000000  ...  0.274705      0.0  0.009899  0.000000   
SRR24874402       0.000000  ...  0.000000      0.0  0.000000  0.000000   
SRR24874404       0.155638  ...  0.000000      0.0  0.037568  0.000000   
SRR24874405       0.000000  ...  0.000000      0.0  0.000000  0.000000   
SRR24874406       0.000000  ...  0.011513      0.0  0.285906  0.007675   

              UCG-010  Varibaculum  Veillonella  Weissella  Xanthomonas  \
Unnamed: 0                                                                
SRR4457864   0.000000          0.0     0.000000        0.0          0.0   
SRR4457873   0.009671          0.0     0.043520        0.0          0.0   
SRR4457876   0.000000          0.0     0.119518        0.0          0.0   
SRR4457877   0.000000          0.0     0.000000        0.0          0.0   
SRR4457878   0.137246          0.0     0.000000        0.0          0.0   
...               ...          ...          ...        ...          ...   
SRR24874401  0.000000          0.0     0.064345        0.0          0.0   
SRR24874402  0.000000          0.0     0.189523        0.0          0.0   
SRR24874404  0.000000          0.0     0.110020        0.0          0.0   
SRR24874405  0.000000          0.0     0.450473        0.0          0.0   
SRR24874406  0.078672          0.0     0.028783        0.0          0.0   

             Unclassified  
Unnamed: 0                 
SRR4457864       3.640935  
SRR4457873       7.630561  
SRR4457876       7.448002  
SRR4457877       6.459400  
SRR4457878       7.